bindings/go/scip/parse.go

```diff
@@ -0,0 +1,149 @@
+package scip
+
+import (
+	"io"
+
+	"google.golang.org/protobuf/encoding/protowire"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// IndexVisitor is a struct of functions rather than an interface since Go
+// doesn't support adding new functions to interfaces with default
+// implementations, so adding new functions here for new fields in
+// the SCIP schema would break clients. Individual functions may be nil.
+type IndexVisitor struct {
+	VisitMetadata       func(*Metadata)
+	VisitDocument       func(*Document)
+	VisitExternalSymbol func(*SymbolInformation)
+}
+
+// ParseStreaming processes an index by incrementally reading input from the io.Reader.
+//
+// Parsing takes place at Document granularity for ease of use.
+func ParseStreaming(r io.Reader, pi IndexVisitor) error {
+	// (field_number << 3) | wire_type
+	// The Index type has less than 15 fields, so the tag
+	// will fit in 1 byte with the varint encoding.
+	tagBuf := make([]byte, 1)
+	// Maximum size of varint is 10 bytes
+	// https://protobuf.dev/programming-guides/encoding/#varints
+	lenBuf := make([]byte, 0, 10)
+	dataBuf := make([]byte, 0, 1024)
+
+	for {
+		numRead, err := r.Read(tagBuf)
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			return errors.Wrapf(err, "failed to read from index reader: %w")
+		}
+		if numRead == 0 {
+			return errors.New("read 0 bytes from index")
+		}
+		fieldNumber, fieldType, errCode := protowire.ConsumeTag(tagBuf)
+		if errCode < 0 {
+			return errors.Wrap(protowire.ParseError(errCode), "failed to consume tag")
+		}
+		switch fieldNumber {
+		// As per scip.proto, all of Metadata, Document and SymbolInformation are sub-messages
+		case metadataFieldNumber, documentsFieldNumber, externalSymbolsFieldNumber:
+			if fieldType != protowire.BytesType {
+				return errors.Newf("expected LEN type tag for %s", indexFieldName(fieldNumber))
+			}
+			lenBuf = lenBuf[:0]
+			dataLen, err := readVarint(r, lenBuf)
+			if err != nil {
+				return errors.Wrapf(err, "failed to read length for %s", indexFieldName(fieldNumber))
+			}
+			if dataLen > uint64(cap(dataBuf)) {
+				dataBuf = make([]byte, dataLen)
+			} else {
+				dataBuf = dataBuf[:0]
+				for i := uint64(0); i < dataLen; i++ {
+					dataBuf = append(dataBuf, 0)
+				}
+			}
+			// Keep going when len == 0 instead of short-circuiting to preserve empty sub-messages
+			if dataLen > 0 {
+				numRead, err := r.Read(dataBuf)
+				if err != nil {
+					return errors.Wrapf(err, "failed to read data for %s", indexFieldName(fieldNumber))
+				}
+				if uint64(numRead) != dataLen {
+					return errors.Newf("expected to read %d bytes based on LEN but read %d bytes", dataLen, numRead)
+				}
+			}
+			if fieldNumber == metadataFieldNumber && pi.VisitMetadata != nil {
```

Comment:
We could generalize some of this using reflection, but I don't think it's that complicated, and the Index structure is probably not going to be changing often enough to warrant the extra complexity of using reflection.